hal/sai_api.h

```diff
@@ -0,0 +1,152 @@
+ * Low level interface to the serial audio interface of a target.
+ *
+ * # Defined behavior
+ * * Supports a subset of the possible configuration space - verified by ::sai_
+ * * Reports a failure (returns false) upon any invocation to an unsupported feature/parameter - verified by ::sai_
+ * * Is able to indicate the currently processed word - verified by ::sai_
+ * # Defined behavior if feature supported
+ * * Is able to change receiver format without interrupting transmitter format - verified by ::sai_
+ * * Is able to change transmitter format without interrupting receiver format - verified by ::sai_
+ * * Is able to change transmitter and receiver format without interrupting the current frame - verified by ::sai_
+ *
+ * @note
+ * A transceiver supporting async rx/tx should be considered as 2 different peripherals :
+ * - one read-only
+ * - one write-only
+ * The first allocated channel may or may not limit the second one's feature.
+ * eg:
+ * In a peripheral that supports async rx/tx but requires format to be the same,
+ * the first allocated instance will set the format and tie the second one to this format.
+ *
+ * @{
+ */
+
+/**
+ * Used to define how transmitter & receiver are related.
+ */
+typedef enum sai_synchronicity_e {
+    sai_synchronicity_async,
+    sai_synchronicity_synced_to_rx,
+    sai_synchronicity_synced_to_tx
+} sai_synchronicity_t;
+
+/**
+ *
+ */
+typedef struct sai_channel_init_s {
+  bool          enable;
+
+  PinName       sd;
+  PinName       bclk;
+  PinName       wclk;
+  bool          is_slave; /**< true if it is slave */
+  bool          internal_bclk; /**< true to use internal bclock */
+  bool          internal_wclk; /**< true to use internal wclock */
+
+  sai_format_t  fmt;
+} sai_channel_init_t;
+
+/**
+ * Used to define communication specs.
+ */
+typedef struct sai_format_s {
+  uint32_t  sample_rate; /**< Typically 44100Hz */
+
+  bool      bclk_polarity; /**< true for Active high */
+  bool      wclk_polarity; /**< true for Active high */
+  bool      ws_delay; /**< true to toggle ws one bit earlier than the frame */
+  uint8_t   ws_length; /**< ws assertion length from 1bclk to word length. */
+
+  uint8_t   frame_length; /**< Frame length in word count [1; 32] */
+  uint32_t  word_mask; /**< Mask on frame for used word (bitfield) */
+  uint8_t   word_length; /**< Word length in bits [1; 32] */
+  uint8_t   data_length; /**< Data length within the word. This must <= to word_length. */
+  bool      lsb_first; /**< true to send lsb first */
+  bool      aligned_left; /**< true to align Left */
+  uint8_t   bit_shift; /**< sample bit shift distance from its alignment point. */
+} sai_format_t;
+
+/** Init structure */
+typedef struct sai_init_s {
+  sai_synchronicity_e sync; // rx/tx independant(aka async), synced to rx, synced to tx
+
+  sai_channel_init_t TX;
+  sai_channel_init_t RX;
+} sai_init_t;
+
+/** Delegated typedef @see target/ ** /object.h */
+typedef struct sai_s sai_t;
```

Comment:
On the FRDM-K66F, it could for example look like this :
```c
struct sai_s {
    I2S_Type *base;
};
```

It may vary depending on the platform/target and the driver used.